rust/aptos-indexer-transaction-stream/src/transaction_stream.rs

```diff
@@ -313,93 +309,150 @@ pub async fn create_fetcher_loop(
-        indexer_grpc_reconnection_timeout_secs,
-        starting_version,
-        request_ending_version,
-        auth_token.clone(),
-        processor_name.to_string(),
-    )
-    .await;
-    let mut connection_id = match response.metadata().get(GRPC_CONNECTION_ID) {
-        Some(connection_id) => connection_id.to_str().unwrap().to_string(),
-        None => "".to_string(),
-    };
-    let mut resp_stream = response.into_inner();
-    info!(
-        processor_name = processor_name,
-        service_type = crate::worker::PROCESSOR_SERVICE_TYPE,
-        stream_address = indexer_grpc_data_service_address.to_string(),
-        connection_id,
-        start_version = starting_version,
-        end_version = request_ending_version,
-        "[Parser] Successfully connected to GRPC stream",
-    );
+    resp_stream: Option<Streaming<TransactionsResponse>>,
+    connection_id: Option<String>,
+    next_version_to_fetch: u64,
+    reconnection_retries: u64,
+    last_fetched_version: i64,
+    fetch_ma: MovingAverage,
+}
 
-    let mut grpc_channel_recv_latency = std::time::Instant::now();
-    let mut next_version_to_fetch = starting_version;
-    let mut reconnection_retries = 0;
-    let mut last_fetched_version = starting_version as i64 - 1;
-    let mut fetch_ma = MovingAverage::new(3000);
-    let mut send_ma = MovingAverage::new(3000);
+pub struct TransactionStreamOutput {
+    pub transactions: Vec<TransactionsPBResponse>,
+    pub should_continue_fetching: bool,
+}
 
-    loop {
-        let is_success = match tokio::time::timeout(
+impl TransactionStream {
+    pub fn new(
+        indexer_grpc_data_service_address: Url,
+        indexer_grpc_http2_ping_interval: Duration,
+        indexer_grpc_http2_ping_timeout: Duration,
+        indexer_grpc_reconnection_timeout_secs: Duration,
+        indexer_grpc_response_item_timeout_secs: Duration,
+        starting_version: u64,
+        request_ending_version: Option<u64>,
+        auth_token: String,
+        processor_name: String,
+        transaction_filter: TransactionFilter,
+        pb_channel_txn_chunk_size: usize,
+    ) -> Self {
+        Self {
+            indexer_grpc_data_service_address,
+            indexer_grpc_http2_ping_interval,
+            indexer_grpc_http2_ping_timeout,
+            indexer_grpc_reconnection_timeout_secs,
             indexer_grpc_response_item_timeout_secs,
-            resp_stream.next(),
+            starting_version,
+            request_ending_version,
+            auth_token,
+            processor_name,
+            transaction_filter,
+            pb_channel_txn_chunk_size,
+            resp_stream: None,
+            connection_id: None,
+            next_version_to_fetch: starting_version,
+            reconnection_retries: 0,
+            last_fetched_version: starting_version as i64 - 1,
+            fetch_ma: MovingAverage::new(3000),
+        }
+    }
+
+    pub async fn init_stream(&mut self) {
```

Comment:
I'm not a huge fan of this `init_stream` approach. It means we need to have fields like `resp_stream` be an `Option` and then unwrap them every time we try to use them, even though there is no case where they should ever be `None` once you've actually initiated the stream. We could make this much more type safe by just doing all this in the `new` method and making `resp_stream` not an `Option` (and therefore eliminating runtime checks like `.expect("[Parser] GRPC stream is not initialized. Did you call init_stream?")`.